tests/test.liq

```diff
@@ -111,6 +111,51 @@ def test.equals(v, v') =
   end
 end
 
+def test.not_equals(first, second) =
+  if
+    first == second
+  then
+    error.raise(
+      error.failure,
+      "expected\r\n
+      #{string.quote(string(first))} == #{string.quote(string(second))}\r\n
+      to be not equal.\r\n"
+    )
+
+    test.fail()
+  end
+end
+
+def test.almost_equals(~places=7, first, second) =
+  diff = abs(float(first) - float(second))
+  if
+    diff * float(pow(10, places)) > 0.5
+  then
+    error.raise(
+      error.failure,
+      "#{string.quote(string(first))} != #{string.quote(string(second))}
+       up to #{places} digits, diff = #{diff}."
+    )
+
+    test.fail()
+  end
+end
+
+def test.not_almost_equals(~places=7, first, second) =
```

Comment:
Let's define this using the previous one so as to not duplicate code logic and potential bugs.

```suggestion
def test.not_almost_equals(~digits=7, first, second) =
  not test.almost_equals(digits=digits, first, second)
```

Reply:
Ha word. Yeah that works. Unfortunately, `%argsof` cannot be used with functions from the same script file currently..

Reply:
I'm not fully understand this suggestion.

I changed
```ruby
def test.not_almost_equals(~places=7, first, second) =
  diff = abs(float(first) - float(second))
  if
    diff * float(pow(10, places)) <= 0.5
  then
    error.raise(
      error.failure,
      "#{string.quote(string(first))} == #{string.quote(string(second))}
       up to #{places} digits."
    )

    test.fail()
  end
end
```
to
```ruby
def test.not_almost_equals(~digits=7, first, second) =
  diff = abs(float(first) - float(second))
  if
    not test.almost_equals(digits=digits, first, second)
  then
    error.raise(
      error.failure,
      "#{string.quote(string(first))} == #{string.quote(string(second))}
       up to #{digits} digits, diff = #{diff}."
    )

    test.fail()
  end
end
```

Reply:
@toots, maybe if I introduce the hidden function that does the math

```ruby
# @flag hidden
def test._almost_equals(~digits=7, first, second) =
  diff = abs(float(first) - float(second))
  if
    diff * float(pow(10, places)) < 0.5
  then
    true.{diff=diff}
  else
    false.{diff=diff}
  end
end
```
then I could reuse it in my code
```ruby
def test.almost_equals(~digits=7, first, second)
  is_almost_equals = test._almost_equals(digits=digits, first, second)
  if not is_almost_equals then
    error.raise(
      error.failure,
      "#{string.quote(string(first))} == #{string.quote(string(second))}
       up to #{digits} digits, diff = #{is_almost_equals.diff}."
    )
    test.fail()
  end
end

def test.not_almost_equals(~digits=7, first, second)
  is_almost_equals = test._almost_equals(digits=digits, first, second)
  if is_almost_equals then
    error.raise(
      error.failure,
      "#{string.quote(string(first))} == #{string.quote(string(second))}
       up to #{digits} digits, diff = #{is_almost_equals.diff}."
    )
    test.fail()
  end
end
```

Reply:
Yes definitely. Also since the test library is not used to generate the doc you don't have to worry to much about hiding it.